pyClarion/numdicts/ops.py

```diff
@@ -56,57 +57,337 @@ def set_by(
 
     return value
 
+
 @register_grad(set_by)
 def _grad_set_by(grads, target, source, *, keyfunc):
 
-    return (grads * 0, sum_by(grads, keyfunc=keyfunc))
+    return (grads * NumDict(default=0), sum_by(grads, keyfunc=keyfunc))
+
+
+@register_op
+def threshold(
+        d: D, *, th: Union[float, int], keep_default: bool = False) -> NumDict:
+    """
+    Return a copy of d containing only values above theshold.
+
+    If the default is below or equal threshold it is set to None in the output, unless
+    keep default is True.
+    """
+
+    mapping = {k: d[k] for k in d if th < d[k]}
+    if d.default is not None:
+        default = d.default if keep_default or th < d.default else None
+    else:  # added this to prevent errors when d.default was none as default was undefined
+        default = None
+    value = NumDict(mapping, default)
+    _kwds = {"th": th, "keep_default": keep_default}
+    record_call(threshold, value, (d,), _kwds)
+    return value
+
+
+@register_grad(threshold)
+def _grad_threshold(grads, d, *, th, keep_default):
+    mapping = {k: (th < d[k]) * grads[k] for k in d}
+    default = grads.default
+    if(d.default == None):
```

Comment:
Please check for None using `d.default is None`. 

I am not sure returning a None default in grads is generally a good idea. Probably better to just return 0. In case the default is not None, it should be subject to the threshold check just like the keys.